File: prova_fortran90/prova3.f90
program prova2
  use omp_lib
  implicit none
  
  integer :: v(0:19), i, ID
  integer :: NUM_THREADS = 4

  !$ call omp_set_num_threads(NUM_THREADS)

  !$OMP PARALLEL
    !$omp critical
    ID = omp_get_thread_num()
    do i=ID,20,+NUM_THREADS
      v(i) = i
      write(*,*) 'v(', i, ')=', v(i), 'ID=', ID
    end do
    !$omp end critical
    !$omp barrier
    !$omp critical
    do i=ID,20,+NUM_THREADS
      v(i) = v(i) + 1
      write(*,*) 'v(', i, ')=', v(i), 'ID=', ID
    end do
    !$omp end critical
    
  !$OMP END PARALLEL


end program prova2

File: prova_C/prova3.c
#include <stdio.h>
#include "omp.h"

#define NUM_THREADS 4
int main(void)
{

  int v[20];
  omp_set_num_threads(NUM_THREADS);
  
  #pragma omp parallel
  {
    int i, ID, nthreads;
    ID = omp_get_thread_num();
    nthreads = omp_get_num_threads();

//    #pragma omp for
    for (i=ID; i<20; i+= nthreads) {
      v[i]=i;
      printf("v[%d] = %d, ID = %d\n",i, v[i], ID);
    }
    #pragma omp barrier
//    #pragma omp for
    for (i=ID; i<20; i+=nthreads) {
      v[i] ++;
      printf("v[%d] = %d, ID = %d\n",i, v[i], ID);
    }
   }  
  
  return 0;
}
